### PR TITLE
Remove stale replication slots on  mirrors

### PR DESCRIPTION
**Pull request by dgkimura**
Stale replication slots can exist on mirrors that were once acting as
primaries. In this case restart_lsn is non-zero value used in past
replication slot setup. Currently there is no scenario using replication
slots on mirrors. The stale replication slots may cause problems. Since
it is referencing a potentially old LSN, if mirror is not connected then
it can collect older WAL logs than necessary. This is only a problem if
mirror is promoted back to primary.

This patch drops internal replication slot on startup as mirror.  There is a known possible race condition in `ReplicationSlotDropIfExists()`, however given that this should only be called once at `StartupXLOG()` this isn't a problem, _yet_..

Co-authored-by: Ashwin Agrawal <aagrawal@pivotal.io>
Co-authored-by: Alexandra Wang <lewang@pivotal.io>

### CODE REVIEW

**Comment by yydzero** [comment]
Is there any readme/doc to describe how replication slot is used in Greenplum?  Some bigger picture will be helpful to review this. Thanks!

**Comment by l-wang** [comment]
> Is there any readme/doc to describe how replication slot is used in Greenplum? Some bigger picture will be helpful to review this. Thanks!

Not as part of this PR but docs for replication slot will be added. @mkiyama correct me if I'm wrong. I think the gist of how Greenplum uses replication slot is that our utilities always create replication slots internally. In Greenplum, replication slots will be created during pg_basebackup and mirror promotion (so we have the slots ready when we do pg_rewind), 'gprecoverseg -F' and 'gprecoverseg -a' will always pass in a --slot option to pg_basebackup and pg_rewind.

File: src/backend/replication/slot.c [hunk 1]
```diff
@@ -1372,3 +1372,29 @@ RestoreSlotFromDisk(const char *name)
 				(errmsg("too many replication slots active before shutdown"),
 				 errhint("Increase max_replication_slots and try again.")));
 }
+
+/*
+ * Permanently drop replication slot identified by the passed in name.
```
Comment (ashwinstar) [inline src/backend/replication/slot.c]: Its worth adding note here mentioning the race condition to warn future potential users for this function. Also, stating why current usage doesn't suffer from it.

File: src/backend/access/transam/xlog.c [hunk 2]
```diff
@@ -6827,6 +6827,11 @@ StartupXLOG(void)
 	 */
 	StartupReplicationSlots();
 
+	if (ArchiveRecoveryRequested)
+	{
```
Comment (ashwinstar) [inline src/backend/access/transam/xlog.c]: we can drop these `{ }`. Also, worth adding comment why this drop is being performed.

**Comment by ashwinstar** [comment]
>> The stale replication slots may cause problems. Since
it is referencing a potentially old LSN, if mirror is not connected then
it can collect older WAL logs than necessary. This is only a problem if
mirror is promoted back to primary.

I don't agree, its problem even while its acting as mirror I think and will continue to collect xlogs.

**Comment by dgkimura** [comment]
>> The stale replication slots may cause problems. Since it is referencing a potentially old LSN, if mirror is not connected then it can collect older WAL logs than necessary. This is only a problem if mirror is promoted back to primary.
>>
> I don't agree, its problem even while its acting as mirror I think and will continue to collect xlogs.

Yes, you are correct.  Thanks for correction!  As you suggested offline, we validated that mirror w/stale replication slot continues to collect WAL by leveraging GUC `create_restartpoint_on_ckpt_record_replay=on` and reducing WAL segment size.

File: src/backend/replication/slot.c (hunk 1 above)
Comment (dgkimura) [reply src/backend/replication/slot.c]: Done. Thanks

File: src/backend/access/transam/xlog.c (hunk 2 above)
Comment (dgkimura) [reply src/backend/access/transam/xlog.c]: Updated.

**Comment by mkiyama** [comment]
> > Is there any readme/doc to describe how replication slot is used in Greenplum? Some bigger picture will be helpful to review this. Thanks!
> 
> @yydzero Not as part of this PR but docs for replication slot will be added. @mkiyama correct me if I'm wrong. I think the gist of how Greenplum uses replication slot is that our utilities always create replication slots internally. In Greenplum, replication slots will be created during pg_basebackup and mirror promotion (so we have the slots ready when we do pg_rewind), 'gprecoverseg -F' and 'gprecoverseg -a' will always pass in a --slot option to pg_basebackup and pg_rewind.

@l-wang Sorry, I wouldn't really know. Since replication slots are an "internal feature," the GPDB user documentation does not have any information on how replication slots are used. The user docs only reference WAL replication general terms.

@yydzero - I believe replication slots are a postgres 9.4 feature. @dgkimura might have better references.

**Comment by ashwinstar** [comment]
Merged https://github.com/greenplum-db/gpdb/commit/fa09dd80f3be01fdc0464977bc6fe54c3d43cb3e

**Comment by pivotal-ning-yu** [comment]
This is really an interesting PR, unlucky I'm late to it.  Would it be possible to keep future PRs open for a longer time so other guys have more chances to share their ideas?

**Comment by ashwinstar** [comment]
On Wed, Jan 23, 2019 at 5:55 PM Ning Yu <notifications@github.com> wrote:

> This is really an interesting PR, unlucky I'm late to it. Would it be
> possible to keep future PRs open for a longer time so other guys have more
> chances to share their ideas?
>
Sorry if was soon, thought was up for more than a day and review comments
were addressed.

Please feel free to provide ideas if any, even if its closed and merged.
Always open to improve the same or re-implement if required.

>

**Comment by pivotal-ning-yu** [comment]
On Thu, Jan 24, 2019 at 2:26 PM Yandong Yao <notifications@github.com>
wrote:

> @pivotal-ning-yu <https://github.com/pivotal-ning-yu> Rob has posted some
> introduction in email group which is helpful.
>
>
> https://mail.google.com/mail/u/0?ik=591c297781&view=om&permmsgid=msg-f%3A1623497266394673005
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__mail.google.com_mail_u_0-3Fik-3D591c297781-26view-3Dom-26permmsgid-3Dmsg-2Df-253A1623497266394673005&d=DwMCaQ&c=lnl9vOaLMzsy2niBC8-h_K-7QJuNJEsFrzdndhuJ3Sw&r=Q6MSdXUs_3UDBMULtNocIw&m=ibnlj8kzv6UcTq0a3U2UR5Oq4scfBIZlv1zQY9GqzBU&s=BXYHc7hc89FO16eQZbhTKWCcpsNr9Uo0qxnaAYNXNlQ&e=>
> .
>

Cool!  Although I can't open your link I believe it's below thread:
https://groups.google.com/a/greenplum.org/d/msg/gpdb-dev/cejpZtm_ZCc/j1n7kI5ZCwAJ

**Comment by yydzero** [comment]
Reading Ashwin's description in above thread,  so we also need to remove stale replication slots on standby.  This PR also covered it already, right?

Before 6, master/standby has different mechanism with primary/mirror. In 6, are there any differences between them, or everything are the same?  If it is latter, guess this PR don't need to differentiate master/standby and primary/mirror.

**Comment by ashwinstar** [comment]
On Tue, Jan 29, 2019 at 6:51 PM Yandong Yao <notifications@github.com>
wrote:

> Reading Ashwin's description in above thread, so we also need to remove
> stale replication slots on standby. This PR also covered it already, right?
>
Yes, it coves as standby and mirrors share the code. But note that
currently for master and standby replication slots are not created and
hence, standby will not have the replication slot anyways. The reason for
master not needing replication slot is anyways for master incremental
recovery is not supported yet, so not required to have replication slot.


> Before 6, master/standby has different mechanism with primary/mirror. In
> 6, are there any differences between them, or everything are the same? If
> it is latter, guess this PR don't need to differentiate master/standby and
> primary/mirror.
>
There still exist some minimal differences between master/standby working
compared to primary/mirror, but majority of the code is the same. But this
PR specifically doesn't need to differentiate between the two.

**Comment by yydzero** [comment]
What is our plan for 'master incremental recovery'?  So surprised to know it is not supported yet.

**Pull request closed without merging**